README.md
# Centralized Logging on AWS Solution

_Note:_ For any relevant information outside the scope of this readme, please refer to the solution landing page and implementation guide.

**Centralized Logging on AWS will be deprecated on March 1, 2024. It is being superseded by the [Centralized Logging with OpenSearch](https://aws.amazon.com/solutions/implementations/centralized-logging-with-opensearch/) (CLOS) solution. After deprecation, all existing deployments will continue to work but the solution will no longer be supported and maintained. If you’re not using GovCloud region deployments, we highly recommend migrating to version 2.0.0 or newer of the CLOS solution by following the steps outlined in the [Centralized Logging on AWS](https://docs.aws.amazon.com/solutions/latest/centralized-logging-on-aws/migrate-to-new-solution.html) implementation guide.**



**[🚀Solution Landing Page](https://aws.amazon.com/solutions/implementations/centralized-logging/)** | **[🚧Feature request](https://github.com/aws-solutions/aws-centralized-logging/issues/new?assignees=&labels=feature-request%2C+enhancement&template=feature_request.md&title=)** | **[🐛Bug Report](https://github.com/aws-solutions/aws-centralized-logging/issues/new?assignees=&labels=bug%2C+triage&template=bug_report.md&title=)** | **[📜Documentation Improvement](https://github.com/aws-solutions/aws-centralized-logging/issues/new?assignees=&labels=document-update&template=documentation_improvements.md&title=)**

## Table of content

- [Solution Overview](#solution-overview)
- [Architecture](#architecture)
- [Installation](#installing-pre-packaged-solution-template)
- [Customization](#customization)
  - [Setup](#setup)
  - [Changes](#changes)
  - [Unit Test](#unit-test)
  - [Build](#build)
  - [Deploy](#deploy)
- [Sample Scenario](#sample-scenario)
- [File Structure](#file-structure)
- [License](#license)

## Solution Overview

Centralized Logging on AWS is a reference implementation that provides a foundation for logging to a centralized account. Customers can leverage the solution to index CloudTrail Logs, CW Logs, VPC Flow Logs on an Amazon OpenSearch Service domain. The logs can then be searched on different fields.

This solution gives you a turnkey environment to begin logging and analyzing your AWS environment and applications. Additionally, if you are looking to

- collect logs from multiple AWS accounts and organizations
- collect logs from multiple regions
- a single pane view for log analysis and visualization

then you can get all this with this 1-click deployment solution.

This solution uses Amazon OpenSearch Service (successor to Amazon Elasticsearch Service) and Kibana, an analytics and visualization platform that is integrated with Amazon OpenSearch Service, that results in a unified view of all the log events.

## Architecture

The Centralized Logging on AWS solution contains the following components: **log ingestion**, **log indexing**, and **visualization**. You must deploy the AWS CloudFormation template in the AWS account where you intend to store your log data.

<img src="./architecture.png" width="750" height="500">

## Installing pre-packaged solution template

- Primary Template: [aws-centralized-logging.template](https://solutions-reference.s3.amazonaws.com/centralized-logging/latest/aws-centralized-logging.template)

- Demo Template: [Demo.template](https://solutions-reference.s3.amazonaws.com/centralized-logging/latest/aws-centralized-logging-demo.template)

## Customization

- Prerequisite: Node.js>=16 | npm >= 8

### Setup

Clone the repository and run the following commands to install dependencies, format and lint as per the project standards

```
npm ci
npm run prettier-format
npm run lint
```

### Changes

You may make any needed change as per your requirement. If you want to customize the Centralized Logging on AWS opinionated defaults, you can modify the [solution manifest file](./source/resources/lib/manifest.json). You can also control sending solution usage metrics to aws-solutions, from the manifest file.

```
"solutionVersion": "%%VERSION%%", #provide a valid value eg. v1.0
"sendMetric": "Yes",
```

Additionally, you can customize the code and add any extension to the solution. Please review our [feature request guidelines](./.github/ISSUE_TEMPLATE/feature_request.md), if you want to submit a PR.

### Unit Test

You can run unit tests with the following command from the root of the project

```
 npm run test
```

### Build

You can build lambda binaries with the following command from the root of the project

```
 npm run build
```

### Deploy

Run the following command from the root of the project. Deploys all the primary solution components needed for Centralized Logging on AWS. **Deploy in Primary Account**

```
cd source/resources
npm ci
```

```
npm run cdk-bootstrap -- --profile <PROFILE_NAME>
npm run cdk-synth
npm run cdk-deploy -- CL-PrimaryStack --parameters AdminEmail=<EMAIL> --parameters SpokeAccounts=<ACCOUNT-ID-1,ACCOUNT-ID-2...> --parameters JumpboxKey=<EC2_KEY_PAIR> --parameters JumpboxDeploy='Yes' --profile <PROFILE_NAME>
```

_Note:_ for PROFILE_NAME, substitute the name of an AWS CLI profile that contains appropriate credentials for deploying in your preferred region.

## Sample Scenario (Enabling CloudWatch logging on Elasticsearch domain)

The default deployment uses opinionated values as setup in [solution manifest file](./source/resources/lib/manifest.json). In this scenario let's say we want to enable CloudWatch logging for ES domain.

You would need to update the **ESDomain** resource in cl-primary-stack.ts as below:

```
 logging: {
        slowSearchLogEnabled: true,
        appLogEnabled: true,
        slowIndexLogEnabled: true,
      },
```

## File structure

Centralized Logging on AWS solution consists of:

- cdk constructs to generate needed resources
- helper for bootstrapping purposes like creating CloudWatch Logs Destinations
- transformer to translate kinesis data stream records into Elasticsearch documents

<pre>
|-config_files                    [ config files for prettier, eslint etc. ]
|-architecture.png                [ solution architecture diagram ]
|-source/
  |dashboard.ndjson               [ sample dashboard for demo ]  
  |run-unit-test.sh               [ script to run unit tests ]
  |-resources
    |-bin/
      |-app.ts                    [ entry point for CDK app ]
    |-__tests__/                  [ unit tests for CDK constructs ] 
    |-lib/
      |-cl-demo-ec2-construct.ts  [ CDK construct for demo web server resource ]
      |-cl-demo-stack.ts          [ CDK construct for demo stack]
      |-cl-jumpbox-construct.ts   [ CDK construct for windows jumpbox resource ]  
      |-cl-primary-stack.ts       [ CDK construct for primary stack and related resources ]
      |-utils.ts                  [ utilities for generic functionalities across CDK constructs ]   
      |-manifest.json             [ manifest file for CDK resources ]
    |-config_files                [ tsconfig, jest.config.js, package.json etc. ]
  |-services/
    |-@aws-solutions/utils/       [ library with generic utility functions for microservice ]
    |-helper/                     [ lambda backed helper custom resource to help with solution launch/update/delete ]
    |-transformer/                [ microservice to translate kinesis records into es documents ]
</pre>

## License

See license [here](./LICENSE.txt)

## Collection of operational metrics

This solution collects anonymized operational metrics to help AWS improve the quality and features of the solution. For more information, including how to disable this capability, please see the [implementation guide](https://docs.aws.amazon.com/solutions/latest/centralized-logging/collection-of-operational-metrics.html).

---

Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

Licensed under the Apache License Version 2.0 (the "License"). You may not use this file except in compliance with the License. A copy of the License is located at

```
http://www.apache.org/licenses/LICENSE-2.0
```

or in the ["license"](./LICENSE.txt) file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and limitations under the License.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [4.0.6] - 2023-10-24

### Changed

- Update dependencies to address [CVE-2023-45133](https://github.com/advisories/GHSA-67hx-6x53-jw92)

## [4.0.5] - 2023-09-18

### Changed

- Update lambda runtime to NodeJS 18
- Update dependencies


## [4.0.4] - 2023-07-07

### Changed

- Bug fix
  - correct batch size for firhose.putRecordBatch
- Upgrade dependencies
- Enforce [imdsV2](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html)

## [4.0.3] - 2023-04-13

### Changed

- Upgrade CDK to version 2
- Upgrade nodejs version for build pipeline to 16
- Upgrade nodejs version for lambda functions to 16
- Include ClusterSize in deployment metrics
- Upgrade aws-sdk version to mitigate [CVE-2023-0842: xml2js vulnerable to prototype pollution](https://nvd.nist.gov/vuln/detail/CVE-2023-0842)

## [4.0.2] - 2022-09-14

### Changed

- updated package versions
- utils logger/metrics moved up to flatten directory hierarchy

## [4.0.1] - 2021-12-05

### Added

- support for '+' in admin email address

### Changed

- aws-cdk updated to 1.132.0
- moved generic helpers to utils library, [logger](./source/services/utils/logger), [metrics](./source/services/utils/metrics)
- CloudWatch Logs destinations created with UUID appended to name for uniqueness

## [4.0.0] - 2020-12-15

### Added

- VPC with 2 isolated & 2 public subnets
- Elasticsearch domain in isolated subnets
- Kinesis Data Stream and Kinesis Firehose for data streaming
- CloudWatch Logs Destination for cross account/region data streaming
- Windows jumpbox for accessing kibana
- Security group for jumpbox
- Security group for ES and Kinesis resources

### Updated

- Elasticsearch V7.7
- Lambda log event transformer
- AWS CDK constructs for IaC

### Removed

- Spoke templates
- Cross account IAM role for Lambda (cross account streaming now uses CloudWatch Logs Destination)

## [3.2.1] - 2020-09-14

### Added

- SNS topic is now encrypted using KMS CMK
- Optional MFA support for Cognito users

### Updated

- Now uses CDK to create deployment templates
- Leverages AWS Solutions Contruct for Lambda/ElasticSearch/Kibana
- Updated to use Amazon Elasticsearch Service v7.7

### Removed

- Demo Access Logging bucket no longer enables versioning
- Removed global egress access from the VPC security group
- Removed all hard-coded logical resource IDs and names to enable multiple stacks to be deployed, such as for testing or migration

## [3.2] - 2019-12-18

### Added

- Backward-compatible to v3.0.0
- Includes all v3.0.1 changes
- Do NOT upgrade from v3.0.1 to v3.2

## [3.0.1] - 2019-11-29

### Added

- Uses SSM Parameters to retrieve the latest HVM x86_64 AMI
- Block public access to 2 buckets created for demo
- CLFullAccessUserRole replaces CognitoAuthorizedRole. It is associated with the Admin group. Initial user is placed in this group.
- CLReadOnlyAccessRole is added. It provides read-only access to users in UserPoolGroupROAccess. This is the default role for Authenticated users in the pool.

### Updated

- Nodejs8.10 to Nodejs12.x Lambda run time.
- Updated license to Apache License version 2.0
- Corrected Master_Role environmental variable in spoke template to MASTER_ROLE
- Updated demo EC2 instance to T3.MICRO
- Updated Nodejs deprecated buffer() to buffer.from()
- Removed python solution-helper from spoke template and replaced with the Nodejs version used by primary.
- Updated NOTICE.txt for 3rd party modules
- Replaced istanbul (deprecated) with nyc
- ElasticSearch version moved to a mapping parameter
- ElasticSearch cluster mappings consolidated under ElasticSearch for clarity/usability
- Tightened security on IAM roles to specific methods and resources

### Removed

- Unreferenced SolutionHelperRole in demo template
- Unreferenced S3 bucket mapping in demo template
- AMIInfo lookup Lambda
- CognitoUnAuthorizedRole / unauthenticated Cognito access

## [0.0.1] - 2019-09-09

### Added

- CHANGELOG template file to fix new pipeline standards

### Updated

- updated buildspec.yml to meet new pipeline build standards
- updated build-s3-dist.sh to meet new pipeline build standards
- updated run-unit-tests.sh for correct references to folders
- updated cloudformation templates to include a bucket for S3 access logs
- updated cloudformation template with correct lambda function environment variable key names


source/services/indexing/index.js
/*******************************************************************************
*  Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved. 
*
*  Licensed under the Apache License Version 2.0 (the "License"). You may not 
*  use this file except in compliance with the License. A copy of the License is 
*  located at                                                           
*
*      http://www.apache.org/licenses/
*
*  or in the "license" file accompanying this file. This file is distributed on  
*  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or 
*  implied. See the License for the specific language governing permissions and  
*  limitations under the License.      
********************************************************************************/
/**
 * @author Solution Builders
 */

'use strict';

let https = require('https');
let zlib = require('zlib');
let crypto = require('crypto');
let AWS = require('aws-sdk');
let moment = require('moment');
const MetricsHelper = require('./lib/metrics-helper.js');
const LOGGER = new(require('./lib/logger'))();

const endpoint = process.env.DOMAIN_ENDPOINT;
const masterRole = process.env.MASTER_ROLE;
const sessionId = process.env.SESSION_ID;
const owner = process.env.OWNER;
const solution = process.env.SOLUTION;
const clusterSize = process.env.CLUSTER_SIZE;
const uuid = process.env.UUID;
const anonymousData = process.env.ANONYMOUS_DATA;

function handler(input, context, callback) {

  let eventText = JSON.stringify(input, null, 2);

  //return callback if environment variable not set
  if (!endpoint || !masterRole || !sessionId || !owner) {
    LOGGER.log('ERROR', `check environment variables endpoint: ${endpoint}, masterRole: ${masterRole}, sessionId: ${sessionId}, owner: ${owner}`);
    return callback('environment variables not defined');
  }

  // Log a message to the console, you can view this text in the Monitoring tab in the Lambda console
  // or in the CloudWatch Logs console
  LOGGER.log('DEBUG', `Received event: ${eventText}`);

  // decode input from base64
  let zippedInput = new Buffer.from(input.awslogs.data, 'base64');

  // decompress the input
  zlib.gunzip(zippedInput, function(error, buffer) {
    if (error) {
      return callback(error);
    }

    // parse the input from JSON
    let awslogsData = JSON.parse(buffer.toString('utf8'));

    // transform the input to Elasticsearch documents
    let elasticsearchBulkData = transform(awslogsData);

    // skip control messages
    if (!elasticsearchBulkData) {
      LOGGER.log('DEBUG', `Received a control message`);
      return callback(null, 'success');
    }
    LOGGER.log('DEBUG', `elasticsearchBulkData: ${elasticsearchBulkData}`);

    // post documents to the Amazon Elasticsearch Service
    post(elasticsearchBulkData, function(error, success,
      statusCode,
      failedItems) {
      LOGGER.log('DEBUG',`Response StatusCode: ${statusCode}`);

      if (error) {
        LOGGER.log('ERROR',`postElasticSearchBulkData Error: ${JSON.stringify(error, null, 2)}`);

        if (failedItems && failedItems.length > 0) {
          LOGGER.log('ERROR', `Failed Items: ${JSON.stringify(failedItems, null, 2)}`);
        }

        return callback(error);

      } else {
        LOGGER.log('INFO',`success: ${JSON.stringify(success)}`);

        if (anonymousData === 'Yes') {

          //send anonymous metrics ONLY if chosen 'Yes'
          /**
           * V38463712 - 08/10/18 - fixing callbacks
           */
          sendMetrics({
            'clusterSize': clusterSize,
            'itemsIndexed': success.successfulItems,
            'totalItemSize': success.totalItemSize
          }, function(err, data) {
            if (err) LOGGER.log('ERROR',`Metrics Status: ${JSON.stringify(err)}`);
            else LOGGER.log('DEBUG',`Metrics Status: ${JSON.stringify(data)}`);
            return callback(null, 'Success');
          });
        } else return callback(null, 'Success');


      }

    });

  });

}

/**
 * Transform CloudWatch Logs stream data for indexing
 * on ES domain
 * @param {JSON} payload - cw log stream data
 */
function transform(payload) {
  if (payload.messageType === 'CONTROL_MESSAGE') {
    return null;
  }

  let bulkRequestBody = '';

  payload.logEvents.forEach(function(logEvent) {
    let timestamp = new Date(1 * logEvent.timestamp);

    // index name format: cwl-YYYY.MM.DD
    let indexName = [
      'cwl-' + timestamp.getUTCFullYear(), // year
      ('0' + (timestamp.getUTCMonth() + 1)).slice(-2), // month
      ('0' + timestamp.getUTCDate()).slice(-2) // day
    ].join('.');

    let source = buildSource(logEvent.message, logEvent.extractedFields);


    /**
     *  V38463712 - 10/09/2018 - Convert field to string
     *  Fix inconsistent CloudTrail attributes & thousands of fields
     */
    if ('requestParameters' in source) source['requestParameters'] = JSON.stringify(source['requestParameters']);
    if ('responseElements' in source) source['responseElements'] = JSON.stringify(source['responseElements']);
    source['@id'] = logEvent.id;
    source['@timestamp'] = new Date(1 * logEvent.timestamp).toISOString();
    source['@message'] = logEvent.message;
    source['@owner'] = payload.owner;
    source['@log_group'] = payload.logGroup;
    source['@log_stream'] = payload.logStream;

    let action = {
      "index": {}
    };
    action.index._index = indexName;
    action.index._type = 'CloudWatchLogs';
    action.index._id = logEvent.id;

    bulkRequestBody += [
      JSON.stringify(action),
      JSON.stringify(source),
    ].join('\n') + '\n';
  });
  LOGGER.log('DEBUG',`bulkRequestBody: ${bulkRequestBody}`);
  return bulkRequestBody;
}

/**
 * Building item for ES indexing
 * @param {String} message - message field from cw event
 * @param {Array} extractedFields - extracted fields from cw event
 */
function buildSource(message, extractedFields) {
  if (extractedFields) {
    let source = {};

    for (let key in extractedFields) {
      if (extractedFields.hasOwnProperty(key) && extractedFields[key]) {
        let value = extractedFields[key];

        if (isNumeric(value)) {
          source[key] = 1 * value;
          continue;
        }

        let jsonSubString = extractJson(value);
        if (jsonSubString !== null) {
          source['$' + key] = JSON.parse(jsonSubString);
        }

        source[key] = value;
      }
    }
    LOGGER.log('DEBUG',`build source output: ${source}`);
    return source;
  }

  let jsonSubString = extractJson(message);
  if (jsonSubString !== null) {
    return JSON.parse(jsonSubString);
  }

  return {};
}

function extractJson(message) {
  let jsonStart = message.indexOf('{');
  if (jsonStart < 0) return null;
  let jsonSubString = message.substring(jsonStart);
  return isValidJson(jsonSubString) ? jsonSubString : null;
}

function isValidJson(message) {
  try {
    JSON.parse(message);
  } catch (e) {
    return false;
  }
  return true;
}

function isNumeric(n) {
  return !isNaN(parseFloat(n)) && isFinite(n);
}

/**
 * Post logs on ES domain.
 * @param {body} body - ES bulk data to be indexed
 * @param {post~callback} callback - The callback that handles the response.
 */
function post(body, callback) {

  LOGGER.log('DEBUG',`endpoint: ${endpoint}`);

  assumeRole(function(err, creds) {
    if (err) {
      LOGGER.log('ERROR',`error in assuming role: ${err}`);
      return callback(err);
    }

    buildRequest(endpoint, body, creds, function(err, requestParams) {
      if (err) {
        LOGGER.log('ERROR',`error in http request: ${err}`);
        return callback(err);
      }

      LOGGER.log('DEBUG',`requestParams: ${requestParams}`);
      let request = https.request(requestParams, function(response) {
        let responseBody = '';
        response.on('data', function(chunk) {
          responseBody += chunk;
        });
        response.on('end', function() {
          let info = JSON.parse(responseBody);
          let failedItems;
          let success;

          LOGGER.log('INFO',`post info: ${JSON.stringify(info,null,2)}`);

          if (response.statusCode >= 200 && response.statusCode <
            299) {
            failedItems = info.items.filter(function(x) {
              return x.index.status >= 300;
            });

            success = {
              "attemptedItems": info.items.length,
              "successfulItems": info.items.length -
                failedItems.length,
              "failedItems": failedItems.length,
              "totalItemSize": requestParams.headers[
                'Content-Length']
            };
          }

          let error = response.statusCode !== 200 || info.errors ===
            true ? {
              "statusCode": response.statusCode,
              "responseBody": responseBody
            } : null;

          LOGGER.log('INFO',`post error: ${JSON.stringify(error,null,2)}`);

          return callback(error, success, response.statusCode,
            failedItems);

        });
      }).on('error', function(e) {
        return callback(e);
      });
      request.end(requestParams.body);
    });

  });

}

/**
 * Assumes role with permissions for ES indexing
 * @param {assumeRole~callback} cb - The callback that handles the response with credentials
 */
function assumeRole(cb) {
  let creds = {
    aws_secret_key: process.env.AWS_SECRET_ACCESS_KEY,
    aws_access_key: process.env.AWS_ACCESS_KEY_ID,
    aws_session_token: process.env.AWS_SESSION_TOKEN
  };

  //assume role in spoke accounts
  if (owner === 'Spoke') {
    //assume role for posting documents on ES Domain
    let sts = new AWS.STS({
      apiVersion: '2011-06-15'
    });
    sts.assumeRole({
      RoleArn: masterRole,
      /* required */
      RoleSessionName: sessionId,
      /* required */
    }, function(err, data) {
      if (err) {
        LOGGER.log('ERROR',`assume role err: ${err}`);
        return cb(err, null);
      } // an error occurred
      else {
        LOGGER.log('DEBUG',`assume role response: ${data}`);
        creds = {
          aws_secret_key: data.Credentials.SecretAccessKey,
          aws_access_key: data.Credentials.AccessKeyId,
          aws_session_token: data.Credentials.SessionToken
        };
        return cb(null, creds);
      }
    });
  } else if (owner === 'Hub') return cb(null, creds);
  else return cb('invalid owner', null);
}

/**
 * Build https request for indexing on ES
 * @param {String} endpoint - ES endpoint for log indexing
 * @param {String} body - - ES bulk data to be indexed
 * @param {Dictionary} creds - AWS credentials for https request
 * @param {buildRequest~callback} cb - The callback that handles the response
 */
function buildRequest(endpoint, body, creds, cb) {
  let endpointParts = endpoint.match(
    /^([^\.]+)\.?([^\.]*)\.?([^\.]*)\.amazonaws\.com$/);
  let region = endpointParts[2];
  let service = endpointParts[3];
  let datetime = (new Date()).toISOString().replace(/[:\-]|\.\d{3}/g, '');
  let date = datetime.substr(0, 8);
  let kDate = hmac('AWS4' + creds.aws_secret_key, date);
  let kRegion = hmac(kDate, region);
  let kService = hmac(kRegion, service);
  let kSigning = hmac(kService, 'aws4_request');

  let request = {
    host: endpoint,
    method: 'POST',
    path: '/_bulk',
    body: body,
    headers: {
      'Content-Type': 'application/json',
      'Host': endpoint,
      'Content-Length': Buffer.byteLength(body),
      'X-Amz-Security-Token': creds.aws_session_token,
      'X-Amz-Date': datetime
    }
  };

  let canonicalHeaders = Object.keys(request.headers)
    .sort(function(a, b) {
      return a.toLowerCase() < b.toLowerCase() ? -1 : 1;
    })
    .map(function(k) {
      return k.toLowerCase() + ':' + request.headers[k];
    })
    .join('\n');

  let signedHeaders = Object.keys(request.headers)
    .map(function(k) {
      return k.toLowerCase();
    })
    .sort()
    .join(';');

  let canonicalString = [
    request.method,
    request.path, '',
    canonicalHeaders, '',
    signedHeaders,
    hash(request.body, 'hex'),
  ].join('\n');

  let credentialString = [date, region, service, 'aws4_request'].join('/');

  let stringToSign = [
    'AWS4-HMAC-SHA256',
    datetime,
    credentialString,
    hash(canonicalString, 'hex')
  ].join('\n');

  request.headers.Authorization = [
    'AWS4-HMAC-SHA256 Credential=' + creds.aws_access_key + '/' +
    credentialString,
    'SignedHeaders=' + signedHeaders,
    'Signature=' + hmac(kSigning, stringToSign, 'hex')
  ].join(', ');

  return cb(null, request);
}

function hmac(key, str, encoding) {
  return crypto.createHmac('sha256', key).update(str, 'utf8').digest(encoding);
}

function hash(str, encoding) {
  return crypto.createHash('sha256').update(str, 'utf8').digest(encoding);
}

function sendMetrics(metricData, cb) {
  let _metricsHelper = new MetricsHelper();

  let _metric = {
    Solution: solution,
    UUID: uuid,
    TimeStamp: moment().utc().format('YYYY-MM-DD HH:mm:ss.S'),
    Data: {
      ClusterSize: metricData.clusterSize,
      ItemsIndexed: metricData.itemsIndexed,
      TotalItemSize: metricData.totalItemSize
    }
  };

  LOGGER.log('DEBUG',`anonymous metric: ${JSON.stringify(_metric)}`);

  _metricsHelper.sendAnonymousMetric(_metric, function(err, data) {
    if (err) {
      let responseData = {
        Error: 'Sending anonymous metric failed'
      };
      LOGGER.log('ERROR',`${[responseData.Error, ':\n', err].join('')}`);
      cb(responseData, null);
    } else {
      let responseStatus = 'SUCCESS';
      let responseData = {
        Success: 'Anonymous metrics sent to AWS'
      };
      cb(null, responseData);
    }
  });

}

module.exports = {
  handler,
  transform,
  buildSource,
  buildRequest,
  assumeRole
};
